File: backend/character_workspace/contracts.py
"""Versioned read contracts for the character workspace."""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, JsonValue


class _ReadModel(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True, str_strip_whitespace=True)


class CharacterWorkspaceErrorCode(str, Enum):
    CHARACTER_NOT_FOUND = "character_not_found"
    TIMELINE_NOT_FOUND = "timeline_not_found"
    TIMELINE_REQUIRED = "timeline_required"
    TIMELINE_CONFLICT = "timeline_conflict"
    CHARACTER_INSTANCE_NOT_FOUND = "character_instance_not_found"
    CHARACTER_INSTANCE_REQUIRED = "character_instance_required"
    CHARACTER_INSTANCE_CONFLICT = "character_instance_conflict"
    INVALID_CURSOR = "invalid_cursor"


class CharacterWorkspaceError(ValueError):
    def __init__(
        self,
        code: CharacterWorkspaceErrorCode,
        message: str,
        *,
        details: dict[str, Any] | None = None,
    ) -> None:
        super().__init__(message)
        self.code = code
        self.details = details or {}


class CharacterRootView(_ReadModel):
    id: UUID
    novel_id: UUID
    name: str
    role_type: str
    description: str = ""
    details: dict[str, JsonValue] = Field(default_factory=dict)
    lifecycle_state: str
    position: int
    version: int
    current_revision_id: UUID | None = None


class TimelineView(_ReadModel):
    id: UUID
    name: str
    timeline_kind: str
    is_primary: bool
    parent_timeline_id: UUID | None = None
    fork_story_sequence: int | None = None


class CharacterInstanceView(_ReadModel):
    id: UUID
    character_id: UUID
    origin_timeline_id: UUID
    continuity_kind: str
    display_label: str = ""
    derived_from_instance_id: UUID | None = None
    lifecycle_state: str
    version: int
    current_revision_id: UUID | None = None
    profile: dict[str, JsonValue] = Field(default_factory=dict)
    profile_schema_version: int | None = None


class CharacterAliasView(_ReadModel):
    id: UUID
    alias: str
    alias_kind: str | None = None
    character_instance_id: UUID | None = None
    timeline_id: UUID | None = None
    identity_layer: str | None = None
    valid_from_sequence: int | None = None
    valid_to_sequence: int | None = None
    lifecycle_state: str


class CharacterRelationshipView(_ReadModel):
    id: UUID
    timeline_id: UUID | None = None
    source_character_id: UUID
    target_character_id: UUID
    source_character_instance_id: UUID | None = None
    target_character_instance_id: UUID | None = None
    directionality: str
    relation_kind: str
    label: str
    description: str = ""
    status: str
    manual_override: bool
    version: int


class ChapterCharacterReference(_ReadModel):
    document_id: UUID
    document_title: str
    document_position: int
    reference_kinds: tuple[Literal["required", "point_of_view"], ...]
    character_instance_id: UUID | None = None
    timeline_id: UUID | None = None


class CharacterVoiceBindingView(_ReadModel):
    binding_id: UUID
    binding_policy: str
    profile_id: UUID | None = None
    voice_version_id: UUID | None = None
    language: str
    version: int


class ProjectedFactView(_ReadModel):
    id: UUID
    fact_type: str
    timeline_id: UUID
    character_id: UUID | None = None
    character_instance_id: UUID | None = None
    relationship_id: UUID | None = None
    dimension: str
    event_kind: str
    predicate: str
    object_text: str
    details: dict[str, JsonValue]
    story_sequence: int | None = None
    source_revision_id: UUID | None = None


class ProjectionConflictView(_ReadModel):
    conflict_key: str
    fact_ids: tuple[UUID, ...]
    reason: str


FactEffectiveState = Literal[
    "current",
    "historical",
    "superseded",
    "source_invalid",
    "batch_reverted",
]
FactHealth = Literal["ok", "conflict", "ambiguous"]


class FactSourceView(_ReadModel):
    document_id: UUID
    document_title: str
    document_position: int
    revision_id: UUID
    revision_is_current: bool
    source_content_hash: str = Field(pattern=r"^[0-9a-f]{64}$")
    source_coordinate: Literal["unicode-codepoint-v1"] = "unicode-codepoint-v1"
    source_start: int | None = Field(default=None, ge=0)
    source_end: int | None = Field(default=None, gt=0)
    source_range_hash: str | None = Field(default=None, pattern=r"^[0-9a-f]{64}$")
    source_excerpt: str = Field(default="", max_length=500)
    source_excerpt_truncated: bool = False
    binding_state: str | None = None
    proposal_item_id: UUID | None = None
    commit_batch_id: UUID | None = None


class ProjectedFactViewV2(ProjectedFactView):
    source_document_id: UUID | None = None
    story_time: dict[str, JsonValue] | None = None
    created_at: datetime
    effective_state: FactEffectiveState
    health: FactHealth
    source: FactSourceView | None = None


class CharacterProjectedStateV2(_ReadModel):
    timeline_id: UUID
    narrative_cutoff: int | None = None
    current_facts: tuple[ProjectedFactViewV2, ...] = ()
    conflicts: tuple[ProjectionConflictView, ...] = ()
    ambiguous_fact_ids: tuple[UUID, ...] = ()


class WritingStateAsOf(_ReadModel):
    timeline_id: UUID
    narrative_cutoff: int | None = None
    story_time: dict[str, JsonValue] | None = None


class WritingStateValue(_ReadModel):
    fact_id: UUID
    object_text: str
    story_sequence: int | None = None
    story_time: dict[str, JsonValue] | None = None
    source: FactSourceView | None = None


class WritingStateSlot(_ReadModel):
    key: Literal[
        "location",
        "goal",
        "health",
        "emotion",
        "identity",
        "knowledge",
        "possession",
        "relationship",
    ]
    label: str
    mode: Literal["single", "multiple"]
    values: tuple[WritingStateValue, ...] = ()
    health: Literal["ok", "conflicted", "ambiguous", "missing"]


class WritingStateRiskSummary(_ReadModel):
    conflict_count: int = Field(ge=0)
    ambiguous_count: int = Field(ge=0)
    invalid_source_count: int = Field(ge=0)


class FactHistorySummary(_ReadModel):
    total: int = Field(ge=0)
    current: int = Field(ge=0)
    historical: int = Field(ge=0)
    superseded: int = Field(ge=0)
    source_invalid: int = Field(ge=0)
    batch_reverted: int = Field(ge=0)


class CharacterWritingState(_ReadModel):
    as_of: WritingStateAsOf
    slots: tuple[WritingStateSlot, ...]
    recent_changes: tuple[ProjectedFactViewV2, ...] = ()
    risk_summary: WritingStateRiskSummary
    history_summary: FactHistorySummary


class CharacterProjectedState(_ReadModel):
    timeline_id: UUID
    narrative_cutoff: int | None = None
    current_facts: tuple[ProjectedFactView, ...] = ()
    conflicts: tuple[ProjectionConflictView, ...] = ()
    ambiguous_fact_ids: tuple[UUID, ...] = ()


class CharacterWorkspaceV1(_ReadModel):
    schema_version: Literal["character-workspace/1"] = "character-workspace/1"
    novel_id: UUID
    character_catalog_version: int
    story_ledger_version: int
    timeline_mode: Literal["single", "multiple"]
    character: CharacterRootView
    selected_timeline: TimelineView
    selected_instance: CharacterInstanceView
    timelines: tuple[TimelineView, ...]
    instances: tuple[CharacterInstanceView, ...]
    aliases: tuple[CharacterAliasView, ...]
    relationships: tuple[CharacterRelationshipView, ...]
    chapter_references: tuple[ChapterCharacterReference, ...]
    voice_binding: CharacterVoiceBindingView | None = None
    projected_state: CharacterProjectedState


class CharacterWorkspaceV2(_ReadModel):
    schema_version: Literal["character-workspace/2"] = "character-workspace/2"
    novel_id: UUID
    character_catalog_version: int
    story_ledger_version: int
    timeline_mode: Literal["single", "multiple"]
    character: CharacterRootView
    selected_timeline: TimelineView
    selected_instance: CharacterInstanceView
    timelines: tuple[TimelineView, ...]
    instances: tuple[CharacterInstanceView, ...]
    aliases: tuple[CharacterAliasView, ...]
    relationships: tuple[CharacterRelationshipView, ...]
    chapter_references: tuple[ChapterCharacterReference, ...]
    voice_binding: CharacterVoiceBindingView | None = None
    projected_state: CharacterProjectedStateV2
    writing_state: CharacterWritingState


class CharacterFactHistoryPage(_ReadModel):
    schema_version: Literal["character-fact-history/1"] = (
        "character-fact-history/1"
    )
    items: tuple[ProjectedFactViewV2, ...]
    next_cursor: str | None = None
    total_summary: FactHistorySummary


class ArchiveImpactReference(_ReadModel):
    reference_type: Literal[
        "active_instance",
        "active_alias",
        "active_relationship",
        "chapter_brief",
        "voice_binding",
        "story_fact",
    ]
    reference_id: UUID
    label: str
    disposition: Literal["requires_review", "preserved_history"]


class CharacterArchiveImpactV1(_ReadModel):
    schema_version: Literal["character-archive-impact/1"] = (
        "character-archive-impact/1"
    )
    novel_id: UUID
    character_id: UUID
    character_name: str
    character_version: int
    already_archived: bool
    requires_confirmation: bool
    current_dependency_count: int
    preserved_history_count: int
    references: tuple[ArchiveImpactReference, ...]

File: frontend/src/characters/contracts.ts
import type {
  IntelligenceBatchRevertCommandV1 as SharedIntelligenceBatchRevertCommandV1,
  StoryFactCorrectionCommandV1 as SharedStoryFactCorrectionCommandV1,
  StoryFactCorrectionResultV1 as SharedStoryFactCorrectionResultV1,
  StoryLedgerFactEffectiveState,
  StoryLedgerFactHealth,
  StoryLedgerBatchImpactPreview,
} from "../story-ledger/contracts";

export type { JsonPrimitive, JsonValue } from "../story-ledger/contracts";
import type { JsonValue } from "../story-ledger/contracts";

export interface CharacterRootView {
  readonly id: string;
  readonly novel_id: string;
  readonly name: string;
  readonly role_type: string;
  readonly description: string;
  readonly details: Readonly<Record<string, JsonValue>>;
  readonly lifecycle_state: string;
  readonly position: number;
  readonly version: number;
  readonly current_revision_id: string | null;
}

export interface TimelineView {
  readonly id: string;
  readonly name: string;
  readonly timeline_kind: string;
  readonly is_primary: boolean;
  readonly parent_timeline_id: string | null;
  readonly fork_story_sequence: number | null;
}

export interface CharacterInstanceView {
  readonly id: string;
  readonly character_id: string;
  readonly origin_timeline_id: string;
  readonly continuity_kind: string;
  readonly display_label: string;
  readonly derived_from_instance_id: string | null;
  readonly lifecycle_state: string;
  readonly version: number;
  readonly current_revision_id: string | null;
  readonly profile: Readonly<Record<string, JsonValue>>;
  readonly profile_schema_version: number | null;
}

export interface CharacterAliasView {
  readonly id: string;
  readonly alias: string;
  readonly alias_kind: string | null;
  readonly timeline_id: string | null;
  readonly character_instance_id: string | null;
  readonly identity_layer: string | null;
  readonly valid_from_sequence: number | null;
  readonly valid_to_sequence: number | null;
  readonly lifecycle_state: string;
}

export interface CharacterRelationshipView {
  readonly id: string;
  readonly timeline_id: string | null;
  readonly source_character_id: string;
  readonly target_character_id: string;
  readonly source_character_instance_id: string | null;
  readonly target_character_instance_id: string | null;
  readonly directionality: string;
  readonly relation_kind: string;
  readonly label: string;
  readonly description: string;
  readonly status: string;
  readonly manual_override: boolean;
  readonly version: number;
}

export interface ChapterCharacterReference {
  readonly document_id: string;
  readonly document_title: string;
  readonly document_position: number;
  readonly reference_kinds: readonly ("required" | "point_of_view")[];
  readonly character_instance_id: string | null;
  readonly timeline_id: string | null;
}

export interface CharacterVoiceBindingView {
  readonly binding_id: string;
  readonly binding_policy: string;
  readonly profile_id: string | null;
  readonly voice_version_id: string | null;
  readonly language: string;
  readonly version: number;
}

interface ProjectedFactCore {
  readonly id: string;
  readonly fact_type: string;
  readonly timeline_id: string;
  readonly character_id: string | null;
  readonly character_instance_id: string | null;
  readonly relationship_id: string | null;
  readonly dimension: string;
  readonly event_kind: string;
  readonly predicate: string;
  readonly object_text: string;
  readonly details: Readonly<Record<string, JsonValue>>;
  readonly story_sequence: number | null;
  readonly source_revision_id: string | null;
}

export interface ProjectionConflictView {
  readonly conflict_key: string;
  readonly fact_ids: readonly string[];
  readonly reason: string;
}

export type CharacterFactEffectiveState = StoryLedgerFactEffectiveState;
export type CharacterFactHealth = StoryLedgerFactHealth;

export interface CharacterFactSourceV2 {
  readonly document_id: string;
  readonly document_title: string;
  readonly document_position: number;
  readonly revision_id: string;
  readonly revision_is_current: boolean;
  readonly source_content_hash: string;
  readonly source_coordinate: "unicode-codepoint-v1";
  readonly source_start: number | null;
  readonly source_end: number | null;
  readonly source_range_hash: string | null;
  readonly source_excerpt: string;
  readonly source_excerpt_truncated: boolean;
  readonly binding_state: string | null;
  readonly proposal_item_id: string | null;
  readonly commit_batch_id: string | null;
}

export interface ProjectedFactViewV2 extends ProjectedFactCore {
  readonly source_document_id: string | null;
  readonly story_time: Readonly<Record<string, JsonValue>> | null;
  readonly created_at: string | null;
  readonly effective_state: CharacterFactEffectiveState;
  readonly health: CharacterFactHealth;
  readonly source: CharacterFactSourceV2 | null;
}

export interface CharacterProjectedStateV2 {
  readonly timeline_id: string;
  readonly narrative_cutoff: number | null;
  readonly current_facts: readonly ProjectedFactViewV2[];
  readonly conflicts: readonly ProjectionConflictView[];
  readonly ambiguous_fact_ids: readonly string[];
}

export interface CharacterWritingStateAsOfV2 {
  readonly timeline_id: string;
  readonly narrative_cutoff: number | null;
  readonly story_time: Readonly<Record<string, JsonValue>> | null;
}

export interface CharacterWritingStateSlotValueV2 {
  readonly fact_id: string;
  readonly object_text: string;
  readonly story_sequence: number | null;
  readonly story_time: Readonly<Record<string, JsonValue>> | null;
  readonly source: CharacterFactSourceV2 | null;
}

export interface CharacterWritingStateSlotV2 {
  readonly key: string;
  readonly label: string;
  readonly mode: string;
  readonly values: readonly CharacterWritingStateSlotValueV2[];
  readonly health: "ok" | "conflicted" | "ambiguous" | "missing";
}

export interface CharacterHistorySummaryV2 {
  readonly total: number;
  readonly current: number;
  readonly historical: number;
  readonly superseded: number;
  readonly source_invalid: number;
  readonly batch_reverted: number;
}

export interface CharacterWritingStateV2 {
  readonly as_of: CharacterWritingStateAsOfV2;
  readonly slots: readonly CharacterWritingStateSlotV2[];
  readonly recent_changes: readonly ProjectedFactViewV2[];
  readonly risk_summary: {
    readonly conflict_count: number;
    readonly ambiguous_count: number;
    readonly invalid_source_count: number;
  };
  readonly history_summary: CharacterHistorySummaryV2;
}

export interface CharacterWorkspaceV2
{
  readonly schema_version: "character-workspace/2";
  readonly novel_id: string;
  readonly character_catalog_version: number;
  readonly story_ledger_version: number;
  readonly timeline_mode: "single" | "multiple";
  readonly character: CharacterRootView;
  readonly selected_timeline: TimelineView;
  readonly selected_instance: CharacterInstanceView;
  readonly timelines: readonly TimelineView[];
  readonly instances: readonly CharacterInstanceView[];
  readonly aliases: readonly CharacterAliasView[];
  readonly relationships: readonly CharacterRelationshipView[];
  readonly chapter_references: readonly ChapterCharacterReference[];
  readonly voice_binding: CharacterVoiceBindingView | null;
  readonly projected_state: CharacterProjectedStateV2;
  readonly writing_state: CharacterWritingStateV2;
}

export interface CharacterRootPatchV2 {
  readonly name: string;
  readonly role_type: string;
  readonly description: string;
  readonly gender: string;
  readonly core_theme: string;
}

export interface CharacterWorkspaceSaveCommandV2 {
  readonly schema_version: "character-workspace-save/2";
  readonly operation_key: string;
  readonly selected_timeline_id: string;
  readonly selected_instance_id: string;
  readonly expected_character_catalog_version: number;
  readonly expected_story_ledger_version: number;
  readonly expected_character_version: number;
  readonly expected_instance_version: number;
  readonly root_patch: CharacterRootPatchV2 | null;
  readonly profile: Readonly<Record<string, JsonValue>> | null;
}

export interface CharacterFactHistoryPageV2 {
  readonly schema_version: "character-fact-history/1";
  readonly items: readonly ProjectedFactViewV2[];
  readonly next_cursor: string | null;
  readonly total_summary: CharacterHistorySummaryV2;
}

export interface CharacterFactHistoryQueryV2 {
  readonly cursor?: string | null;
  readonly limit?: number;
  readonly effective_state?: CharacterFactEffectiveState | "all";
  readonly health?: CharacterFactHealth | "all";
  readonly dimension?: string | null;
  readonly source_document_id?: string | null;
}

export type StoryFactCorrectionCommandV1 = SharedStoryFactCorrectionCommandV1;
export type StoryFactCorrectionResultV1 = SharedStoryFactCorrectionResultV1;
export type CharacterBatchRevertImpact = StoryLedgerBatchImpactPreview;
export type IntelligenceBatchRevertCommandV1 = SharedIntelligenceBatchRevertCommandV1;

export interface CharacterWorkspaceSelection {
  readonly timelineId: string;
  readonly instanceId: string;
}

export interface CharacterWorkspaceFieldErrors {
  readonly [field: string]: string;
}

export interface CharacterWorkspaceActionError {
  readonly code: string;
  readonly message: string;
  readonly field_errors?: CharacterWorkspaceFieldErrors;
  readonly current_workspace?: CharacterWorkspaceV2;
}

export interface CharacterWorkspaceVoiceSlotProps {
  readonly novelId: string;
  readonly characterId: string;
  readonly characterName: string;
  readonly binding: CharacterVoiceBindingView | null;
}
